### PR TITLE
Card payment processor implementation

### PR DESCRIPTION
**Pull request by EvgeniiZaitsevCW**
### Main changes:
1.  The card payment processor contract has been rewritten to Solidity 0.8.x based on the code from 
[brlc-token](https://github.com/cloudwalk/brlc-token) repo. Version 0.8.15 was used for compilation and testing.
2.  Instead of using the whitelist and pauser the roles have been introduced based on the [OpenZeppelin access control contracts](https://docs.openzeppelin.com/contracts/4.x/access-control). There are three roles now: the owner, the pauser, the executor.
3. The tests have been renovated too.

### Testing and coverage
The updated and newly developed tests have been successfully run on the following blockchains:
1. Internal Hardhat net.
2. Substrate with Frontier layer built from the [cloudwalk-network](https://github.com/cloudwalk/cloudwalk-network) repo, branch [main](https://github.com/cloudwalk/cloudwalk-network), commit [350445547016ce83c5ce781ac8871194439b0e4f](https://github.com/cloudwalk/cloudwalk-network/commit/350445547016ce83c5ce781ac8871194439b0e4f).

_NOTE 1_: When running a local node of Substrate with Frontier layer for testing use the node in the archive mode like in the following command:
```bash
./target/release/cloudwalk-network-node --dev --tmp --pruning archive
```

Coverage:
coverage_CPP_0 8 0

### CODE REVIEW

File: test-utils/misc.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      `The length of the base string if greater than allowed maximum length. ` +
+      `The base string: '${baseString}'. ` +
+      `The target byte length: '${byteLength}'`
+    );
+  }
+  if (!/^[0-9a-fA-F]+$/.test(baseString)) {
+    throw new Error(
+      `Creating of bytes16 string failed. ` +
+      `The base string content is incorrect. ` +
+      `The base string: '${baseString}'`
```
Comment (cwteles) [inline test-utils/misc.ts:24]: Codacy has a fix for the issue: [Insert `··`](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
        `The base string: '${baseString}'`
```

File: test/CardPaymentProcessorUpgradeable.test.ts [hunk 2]
```diff
@@ -0,0 +1,1461 @@
...
+  }
+
+  async function checkPaymentStructuresOnBlockchain(payments: TestPayment[]) {
+    for (let i = 0; i < payments.length; ++i) {
+      const expectedPayment: TestPayment = payments[i];
+      const actualPayment = await cardPaymentProcessor.paymentFor(
+        createBytesString(expectedPayment.authorizationId, BYTES16_LENGTH)
+      );
+      checkEquality(actualPayment, expectedPayment, i);
+      if (!!expectedPayment.parentTxHash) {
```
Comment (cwteles) [inline test/CardPaymentProcessorUpgradeable.test.ts:218]: Codacy has a fix for the issue: [Redundant double negation.](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
      if (expectedPayment.parentTxHash) {
```

File: test/CardPaymentProcessorUpgradeable.test.ts [hunk 3]
```diff
@@ -0,0 +1,1461 @@
...
+
+    return balancesPerAccount;
+  }
+
+  function defineExpectedCardPaymentProcessorState(payments: TestPayment[]): CardPaymentProcessorState {
+    const tokenBalance: number = countNumberArrayTotal(
+      payments.map(
+        function (payment: TestPayment): number {
+          return payment.status == PaymentStatus.Uncleared || payment.status == PaymentStatus.Cleared
+            ? payment.amount
```
Comment (cwteles) [inline test/CardPaymentProcessorUpgradeable.test.ts:179]: Codacy has a fix for the issue: [Delete `··`](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
          ? payment.amount
```

File: test/CardPaymentProcessorUpgradeable.test.ts [hunk 4]
```diff
@@ -0,0 +1,1461 @@
...
+    });
+
+    return balancesPerAccount;
+  }
+
+  function defineExpectedCardPaymentProcessorState(payments: TestPayment[]): CardPaymentProcessorState {
+    const tokenBalance: number = countNumberArrayTotal(
+      payments.map(
+        function (payment: TestPayment): number {
+          return payment.status == PaymentStatus.Uncleared || payment.status == PaymentStatus.Cleared
```
Comment (cwteles) [inline test/CardPaymentProcessorUpgradeable.test.ts:178]: Codacy has a fix for the issue: [Replace `··········return·payment.status·==·PaymentStatus.Uncleared·||` with `········return·payment.status·==·PaymentStatus.Uncleared·||⏎·········`](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
        return payment.status == PaymentStatus.Uncleared ||
```

File: test/CardPaymentProcessorUpgradeable.test.ts [hunk 5]
```diff
@@ -0,0 +1,1461 @@
...
+  let user1: SignerWithAddress;
+  let user2: SignerWithAddress;
+  let ownerRole: string;
+  let pauserRole: string;
+  let executorRole: string;
+
+  async function setUpContractsForPayments(payments: TestPayment[]) {
+    for (let payment of payments) {
+      await proveTx(tokenMock.mint(payment.account.address, payment.amount));
+      const allowance: BigNumber = await tokenMock.allowance(payment.account.address, cardPaymentProcessor.address);
```
Comment (cwteles) [inline test/CardPaymentProcessorUpgradeable.test.ts:125]: Codacy has a fix for the issue: [Replace `payment.account.address,·cardPaymentProcessor.address` with `⏎········payment.account.address,⏎········cardPaymentProcessor.address⏎······`](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
      const allowance: BigNumber = await tokenMock.allowance(
```

File: test/CardPaymentProcessorUpgradeable.test.ts [hunk 6]
```diff
@@ -0,0 +1,1461 @@
...
+          ethers.constants.MaxUint256
+        ));
+      }
+    }
+  }
+
+  async function makePayments(payments: TestPayment[]) {
+    for (let payment of payments) {
+      await proveTx(cardPaymentProcessor.connect(payment.account).makePayment(
+        payment.amount,
```
Comment (cwteles) [inline test/CardPaymentProcessorUpgradeable.test.ts:138]: Codacy has a fix for the issue: [Replace `········` with `············`](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
            payment.amount,
```

File: test/CardPaymentProcessorUpgradeable.test.ts [hunk 7]
```diff
@@ -0,0 +1,1461 @@
...
+    await proveTx(cardPaymentProcessor.grantRole(executorRole, account.address));
+  }
+
+  async function clearPayments(payments: TestPayment[], adminAccount: SignerWithAddress) {
+    const authorizationIds: string[] = [];
+    payments.forEach((payment: TestPayment) => {
+      authorizationIds.push(createBytesString(payment.authorizationId, BYTES16_LENGTH));
+      payment.status = PaymentStatus.Cleared;
+    });
+    await proveTx(cardPaymentProcessor.connect(adminAccount).clearPayments(authorizationIds));
```
Comment (cwteles) [inline test/CardPaymentProcessorUpgradeable.test.ts:156]: Codacy has a fix for the issue: [Replace `cardPaymentProcessor.connect(adminAccount).clearPayments(authorizationIds)` with `⏎······cardPaymentProcessor.connect(adminAccount).clearPayments(authorizationIds)⏎····`](https://app.codacy.com/gh/cloudwalk/brlc-periphery/pullRequest?prid=9942053)
```suggestion
    await proveTx(
```

**Review by igorsenych-cw — APPROVED** [review]

**Pull request merged**